File: image-manipulation-js/pixels.js
function getPixelIndices(width, x, y) {
  const rIdx = y * (width * 4) + x * 4
  const gIdx = rIdx + 1
  const bIdx = rIdx + 2
  const aIdx = rIdx + 3

  return [ rIdx, gIdx, bIdx, aIdx ]
}

export function getPixel(imgData, x, y) {
  const [ rIdx, gIdx, bIdx, aIdx ] = getPixelIndices(imgData.width, x, y)

  return {
      r: imgData.data[rIdx],
      g: imgData.data[gIdx],
      b: imgData.data[bIdx],
      a: imgData.data[aIdx]
  }
}

export function setPixel(imgData, x, y, { r, g, b, a }) {
  const [ rIdx, gIdx, bIdx, aIdx ] = getPixelIndices(imgData.width, x, y)

  imgData.data[rIdx] = r
  imgData.data[gIdx] = g
  imgData.data[bIdx] = b
  imgData.data[aIdx] = a
}

File: image-manipulation-zig/pixels.zig
pub fn get_pixel(ImageData imgData, r_idx: usize, c_idx: usize) RgbPixel {
    const red_idx = (r_idx * imgData.n_cols + c_idx) * 4;

    return RgbPixel{ .r = imgData.data[red_idx], .g = imgData.data[red_idx + 1], .b = imgData.data[red_idx + 2], .a = imgData.data[red_idx + 3] };
}

pub fn set_pixel(ImageData imgData, r_idx: usize, c_idx: usize, r: u8, g: u8, b: u8, a: u8) void {
    const red_idx: usize = (r_idx * imgData.n_cols + c_idx) * 4;
    imgData.data[red_idx] = r;
    imgData.data[red_idx + 1] = g;
    imgData.data[red_idx + 2] = b;
    imgData.data[red_idx + 3] = a;
}
